File: sleep_calculator.js
// sleep_calculator.js
const fs = require('fs').promises;
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const question = (prompt) => new Promise(resolve => rl.question(prompt, resolve));

const CYCLE_MINUTES = 90;

class SleepEntry {
    constructor(id, date, bedtime, wakeup, duration, quality) {
        this.id = id;
        this.date = date;
        this.bedtime = bedtime;
        this.wakeup = wakeup;
        this.duration = duration;
        this.quality = quality;
    }
}

class SleepCalculator {
    constructor() {
        this.entries = [];
        this.nextId = 1;
    }

    parseTime(timeStr) {
        const parts = timeStr.split(':');
        if (parts.length !== 2) throw new Error('Неверный формат, используйте ЧЧ:ММ');
        const h = parseInt(parts[0]);
        const m = parseInt(parts[1]);
        if (isNaN(h) || isNaN(m) || h < 0 || h > 23 || m < 0 || m > 59)
            throw new Error('Неверные часы или минуты');
        return { h, m };
    }

    addMinutesToTime(timeStr, minutes) {
        const { h, m } = this.parseTime(timeStr);
        const dt = new Date(2000, 0, 1, h, m);
        dt.setMinutes(dt.getMinutes() + minutes);
        return String(dt.getHours()).padStart(2, '0') + ':' + String(dt.getMinutes()).padStart(2, '0');
    }

    calculateWakeup(bedtime, cycles = 5) {
        if (cycles < 1 || cycles > 10) throw new Error('Количество циклов должно быть от 1 до 10');
        return this.addMinutesToTime(bedtime, cycles * CYCLE_MINUTES);
    }

    calculateBedtime(wakeup, cycles = 5) {
        if (cycles < 1 || cycles > 10) throw new Error('Количество циклов должно быть от 1 до 10');
        const total = cycles * CYCLE_MINUTES;
        const { h, m } = this.parseTime(wakeup);
        const dt = new Date(2000, 0, 1, h, m);
        dt.setMinutes(dt.getMinutes() - total);
        return String(dt.getHours()).padStart(2, '0') + ':' + String(dt.getMinutes()).padStart(2, '0');
    }

    addEntry(bedtime, wakeup, quality) {
        if (quality < 1 || quality > 5) throw new Error('Оценка качества должна быть от 1 до 5');
        const t1 = this.parseTime(bedtime);
        const t2 = this.parseTime(wakeup);
        let dt1 = new Date(2000, 0, 1, t1.h, t1.m);
        let dt2 = new Date(2000, 0, 1, t2.h, t2.m);
        if (dt2 <= dt1) dt2.setDate(dt2.getDate() + 1);
        const duration = Math.round((dt2 - dt1) / 60000);
        const entry = new SleepEntry(
            this.nextId,
            new Date().toISOString().slice(0,10),
            bedtime,
            wakeup,
            duration,
            quality
        );
        this.entries.push(entry);
        this.nextId++;
        return entry;
    }

    findEntry(id) {
        return this.entries.find(e => e.id === id);
    }

    deleteEntry(id) {
        const index = this.entries.findIndex(e => e.id === id);
        if (index === -1) return false;
        this.entries.splice(index, 1);
        return true;
    }

    getStats() {
        const total = this.entries.length;
        if (total === 0) return { total: 0, avgDuration: null, avgQuality: null };
        const sumDur = this.entries.reduce((s, e) => s + e.duration, 0);
        const sumQual = this.entries.reduce((s, e) => s + e.quality, 0);
        return {
            total,
            avgDuration: sumDur / total,
            avgQuality: sumQual / total
        };
    }

    async saveToFile(filename = 'sleep_data.json') {
        const data = { entries: this.entries };
        await fs.writeFile(filename, JSON.stringify(data, null, 2));
    }

    async loadFromFile(filename = 'sleep_data.json') {
        try {
            const data = await fs.readFile(filename, 'utf8');
            const parsed = JSON.parse(data);
            this.entries = parsed.entries.map(e => Object.assign(new SleepEntry(0), e));
            this.nextId = this.entries.reduce((max, e) => Math.max(max, e.id), 0) + 1;
        } catch (err) {
            if (err.code !== 'ENOENT') throw err;
        }
    }
}

function printEntry(entry) {
    const stars = '⭐'.repeat(entry.quality) + '☆'.repeat(5 - entry.quality);
    console.log(`#${entry.id} - ${entry.date} | Лёг: ${entry.bedtime} | Проснулся: ${entry.wakeup}`);
    console.log(`   Продолжительность: ${entry.duration} мин (${Math.floor(entry.duration/60)}ч ${entry.duration%60}м)`);
    console.log(`   Качество: ${stars} (${entry.quality}/5)`);
}

async function main() {
    const calc = new SleepCalculator();
    await calc.loadFromFile();

    while (true) {
        console.log('\n===== КАЛЬКУЛЯТОР СНА (JavaScript) =====');
        console.log('1. Рассчитать время пробуждения');
        console.log('2. Рассчитать время засыпания');
        console.log('3. Добавить запись о сне');
        console.log('4. Показать все записи');
        console.log('5. Показать статистику');
        console.log('6. Удалить запись');
        console.log('7. Сохранить в файл');
        console.log('8. Загрузить из файла');
        console.log('0. Выход');
        const choice = await question('Выберите действие: ');

        if (choice === '0') break;

        switch (choice) {
            case '1': {
                const bedtime = await question('Введите время засыпания (ЧЧ:ММ): ');
                let cyclesInput = await question('Введите количество циклов (по умолчанию 5): ');
                let cycles = cyclesInput.trim() === '' ? 5 : parseInt(cyclesInput);
                try {
                    const wakeup = calc.calculateWakeup(bedtime, cycles);
                    console.log(`Вы должны проснуться в: ${wakeup}`);
                } catch (err) {
                    console.log('Ошибка:', err.message);
                }
                break;
            }
            case '2': {
                const wakeup = await question('Введите время пробуждения (ЧЧ:ММ): ');
                let cyclesInput = await question('Введите количество циклов (по умолчанию 5): ');
                let cycles = cyclesInput.trim() === '' ? 5 : parseInt(cyclesInput);
                try {
                    const bedtime = calc.calculateBedtime(wakeup, cycles);
                    console.log(`Вам следует лечь спать в: ${bedtime}`);
                } catch (err) {
                    console.log('Ошибка:', err.message);
                }
                break;
            }
            case '3': {
                const bedtime = await question('Время засыпания (ЧЧ:ММ): ');
                const wakeup = await question('Время пробуждения (ЧЧ:ММ): ');
                const quality = parseInt(await question('Оценка качества (1-5): '));
                try {
                    const entry = calc.addEntry(bedtime, wakeup, quality);
                    console.log(`Запись добавлена с ID ${entry.id}`);
                } catch (err) {
                    console.log('Ошибка:', err.message);
                }
                break;
            }
            case '4':
                if (calc.entries.length === 0) {
                    console.log('Записей нет.');
                } else {
                    calc.entries.forEach(printEntry);
                }
                break;
            case '5': {
                const stats = calc.getStats();
                if (stats.total === 0) {
                    console.log('Нет записей для статистики.');
                } else {
                    console.log('\n=== СТАТИСТИКА ===');
                    console.log(`Всего записей: ${stats.total}`);
                    console.log(`Средняя продолжительность: ${stats.avgDuration.toFixed(1)} мин (${Math.floor(stats.avgDuration/60)}ч ${Math.round(stats.avgDuration%60)}м)`);
                    console.log(`Средняя оценка качества: ${stats.avgQuality.toFixed(2)}/5`);
                }
                break;
            }
            case '6': {
                const id = parseInt(await question('Введите ID записи для удаления: '));
                if (calc.deleteEntry(id)) {
                    console.log('Запись удалена.');
                } else {
                    console.log('Запись не найдена.');
                }
                break;
            }
            case '7':
                try {
                    await calc.saveToFile();
                    console.log('Сохранено.');
                } catch (err) {
                    console.log('Ошибка сохранения:', err.message);
                }
                break;
            case '8':
                try {
                    await calc.loadFromFile();
                    console.log('Загружено.');
                } catch (err) {
                    console.log('Ошибка загрузки:', err.message);
                }
                break;
            default:
                console.log('Неизвестная команда.');
        }
    }
    rl.close();
}

main().catch(console.error);

File: sleep_calculator.go
// sleep_calculator.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

const cycleMinutes = 90

type SleepEntry struct {
	ID       int    `json:"id"`
	Date     string `json:"date"`
	Bedtime  string `json:"bedtime"`  // HH:MM
	Wakeup   string `json:"wakeup"`   // HH:MM
	Duration int    `json:"duration"` // minutes
	Quality  int    `json:"quality"`  // 1-5
}

type SleepData struct {
	Entries []SleepEntry `json:"entries"`
}

type SleepCalculator struct {
	entries []SleepEntry
	nextID  int
}

func NewSleepCalculator() *SleepCalculator {
	return &SleepCalculator{
		entries: []SleepEntry{},
		nextID:  1,
	}
}

func (c *SleepCalculator) parseTime(timeStr string) (int, int, error) {
	parts := strings.Split(timeStr, ":")
	if len(parts) != 2 {
		return 0, 0, fmt.Errorf("неверный формат, используйте ЧЧ:ММ")
	}
	h, err := strconv.Atoi(parts[0])
	if err != nil || h < 0 || h > 23 {
		return 0, 0, fmt.Errorf("неверные часы")
	}
	m, err := strconv.Atoi(parts[1])
	if err != nil || m < 0 || m > 59 {
		return 0, 0, fmt.Errorf("неверные минуты")
	}
	return h, m, nil
}

func (c *SleepCalculator) addMinutesToTime(timeStr string, minutes int) (string, error) {
	h, m, err := c.parseTime(timeStr)
	if err != nil {
		return "", err
	}
	t := time.Date(2000, 1, 1, h, m, 0, 0, time.UTC)
	t = t.Add(time.Duration(minutes) * time.Minute)
	return t.Format("15:04"), nil
}

func (c *SleepCalculator) CalculateWakeup(bedtime string, cycles int) (string, error) {
	if cycles < 1 || cycles > 10 {
		return "", fmt.Errorf("количество циклов должно быть от 1 до 10")
	}
	total := cycles * cycleMinutes
	return c.addMinutesToTime(bedtime, total)
}

func (c *SleepCalculator) CalculateBedtime(wakeup string, cycles int) (string, error) {
	if cycles < 1 || cycles > 10 {
		return "", fmt.Errorf("количество циклов должно быть от 1 до 10")
	}
	total := cycles * cycleMinutes
	h, m, err := c.parseTime(wakeup)
	if err != nil {
		return "", err
	}
	t := time.Date(2000, 1, 1, h, m, 0, 0, time.UTC)
	t = t.Add(-time.Duration(total) * time.Minute)
	return t.Format("15:04"), nil
}

func (c *SleepCalculator) AddEntry(bedtime, wakeup string, quality int) (SleepEntry, error) {
	if quality < 1 || quality > 5 {
		return SleepEntry{}, fmt.Errorf("оценка качества должна быть от 1 до 5")
	}
	h1, m1, err := c.parseTime(bedtime)
	if err != nil {
		return SleepEntry{}, err
	}
	h2, m2, err := c.parseTime(wakeup)
	if err != nil {
		return SleepEntry{}, err
	}
	t1 := time.Date(2000, 1, 1, h1, m1, 0, 0, time.UTC)
	t2 := time.Date(2000, 1, 1, h2, m2, 0, 0, time.UTC)
	if t2.Before(t1) || t2.Equal(t1) {
		t2 = t2.Add(24 * time.Hour)
	}
	duration := int(t2.Sub(t1).Minutes())
	entry := SleepEntry{
		ID:       c.nextID,
		Date:     time.Now().Format("2006-01-02"),
		Bedtime:  bedtime,
		Wakeup:   wakeup,
		Duration: duration,
		Quality:  quality,
	}
	c.entries = append(c.entries, entry)
	c.nextID++
	return entry, nil
}

func (c *SleepCalculator) FindEntry(id int) *SleepEntry {
	for i := range c.entries {
		if c.entries[i].ID == id {
			return &c.entries[i]
		}
	}
	return nil
}

func (c *SleepCalculator) DeleteEntry(id int) bool {
	for i, e := range c.entries {
		if e.ID == id {
			c.entries = append(c.entries[:i], c.entries[i+1:]...)
			return true
		}
	}
	return false
}

func (c *SleepCalculator) Stats() map[string]interface{} {
	total := len(c.entries)
	if total == 0 {
		return map[string]interface{}{"total": 0, "avg_duration": nil, "avg_quality": nil}
	}
	sumDur := 0
	sumQual := 0
	for _, e := range c.entries {
		sumDur += e.Duration
		sumQual += e.Quality
	}
	return map[string]interface{}{
		"total":        total,
		"avg_duration": float64(sumDur) / float64(total),
		"avg_quality":  float64(sumQual) / float64(total),
	}
}

func (c *SleepCalculator) SaveToFile(filename string) error {
	data := SleepData{Entries: c.entries}
	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filename, jsonData, 0644)
}

func (c *SleepCalculator) LoadFromFile(filename string) error {
	data, err := os.ReadFile(filename)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	var sd SleepData
	if err := json.Unmarshal(data, &sd); err != nil {
		return err
	}
	c.entries = sd.Entries
	for _, e := range c.entries {
		if e.ID >= c.nextID {
			c.nextID = e.ID + 1
		}
	}
	return nil
}

func readString(prompt string) string {
	fmt.Print(prompt)
	reader := bufio.NewReader(os.Stdin)
	input, _ := reader.ReadString('\n')
	return strings.TrimSpace(input)
}

func readInt(prompt string) int {
	for {
		input := readString(prompt)
		if val, err := strconv.Atoi(input); err == nil {
			return val
		}
		fmt.Println("Введите число.")
	}
}

func printEntry(entry SleepEntry) {
	stars := strings.Repeat("⭐", entry.Quality) + strings.Repeat("☆", 5-entry.Quality)
	fmt.Printf("#%d - %s | Лёг: %s | Проснулся: %s\n", entry.ID, entry.Date, entry.Bedtime, entry.Wakeup)
	fmt.Printf("   Продолжительность: %d мин (%dч %dм)\n", entry.Duration, entry.Duration/60, entry.Duration%60)
	fmt.Printf("   Качество: %s (%d/5)\n", stars, entry.Quality)
}

func main() {
	calc := NewSleepCalculator()
	if err := calc.LoadFromFile("sleep_data.json"); err != nil {
		fmt.Println("Ошибка загрузки:", err)
	}

	for {
		fmt.Println("\n===== КАЛЬКУЛЯТОР СНА (Go) =====")
		fmt.Println("1. Рассчитать время пробуждения")
		fmt.Println("2. Рассчитать время засыпания")
		fmt.Println("3. Добавить запись о сне")
		fmt.Println("4. Показать все записи")
		fmt.Println("5. Показать статистику")
		fmt.Println("6. Удалить запись")
		fmt.Println("7. Сохранить в файл")
		fmt.Println("8. Загрузить из файла")
		fmt.Println("0. Выход")
		choice := readString("Выберите действие: ")

		switch choice {
		case "0":
			return
		case "1":
			bedtime := readString("Введите время засыпания (ЧЧ:ММ): ")
			cycles := 5
			if input := readString("Введите количество циклов (по умолчанию 5): "); input != "" {
				cycles, _ = strconv.Atoi(input)
			}
			if wakeup, err := calc.CalculateWakeup(bedtime, cycles); err != nil {
				fmt.Println("Ошибка:", err)
			} else {
				fmt.Printf("Вы должны проснуться в: %s\n", wakeup)
			}
		case "2":
			wakeup := readString("Введите время пробуждения (ЧЧ:ММ): ")
			cycles := 5
			if input := readString("Введите количество циклов (по умолчанию 5): "); input != "" {
				cycles, _ = strconv.Atoi(input)
			}
			if bedtime, err := calc.CalculateBedtime(wakeup, cycles); err != nil {
				fmt.Println("Ошибка:", err)
			} else {
				fmt.Printf("Вам следует лечь спать в: %s\n", bedtime)
			}
		case "3":
			bedtime := readString("Время засыпания (ЧЧ:ММ): ")
			wakeup := readString("Время пробуждения (ЧЧ:ММ): ")
			quality := readInt("Оценка качества (1-5): ")
			entry, err := calc.AddEntry(bedtime, wakeup, quality)
			if err != nil {
				fmt.Println("Ошибка:", err)
			} else {
				fmt.Printf("Запись добавлена с ID %d\n", entry.ID)
			}
		case "4":
			if len(calc.entries) == 0 {
				fmt.Println("Записей нет.")
			} else {
				for _, e := range calc.entries {
					printEntry(e)
				}
			}
		case "5":
			stats := calc.Stats()
			if stats["total"] == 0 {
				fmt.Println("Нет записей для статистики.")
			} else {
				fmt.Println("\n=== СТАТИСТИКА ===")
				fmt.Printf("Всего записей: %d\n", stats["total"])
				fmt.Printf("Средняя продолжительность: %.1f мин (%.0fч %.0fм)\n",
					stats["avg_duration"], stats["avg_duration"].(float64)/60, math.Mod(stats["avg_duration"].(float64), 60))
				fmt.Printf("Средняя оценка качества: %.2f/5\n", stats["avg_quality"])
			}
		case "6":
			id := readInt("Введите ID записи для удаления: ")
			if calc.DeleteEntry(id) {
				fmt.Println("Запись удалена.")
			} else {
				fmt.Println("Запись не найдена.")
			}
		case "7":
			if err := calc.SaveToFile("sleep_data.json"); err != nil {
				fmt.Println("Ошибка сохранения:", err)
			} else {
				fmt.Println("Сохранено.")
			}
		case "8":
			if err := calc.LoadFromFile("sleep_data.json"); err != nil {
				fmt.Println("Ошибка загрузки:", err)
			} else {
				fmt.Println("Загружено.")
			}
		default:
			fmt.Println("Неизвестная команда.")
		}
	}
}
